systems/controllers/osc/osc_tracking_data.cc

```diff
@@ -0,0 +1,387 @@
+    const RigidBodyTree<double>* tree_wo_spr) : TaskSpaceTrackingData(name, n_r,
+          K_p, K_d, W, tree_w_spr, tree_wo_spr) {
+}
+
+void TransTaskSpaceTrackingData::AddPointToTrack(std::string body_name,
+    Vector3d pt_on_body) {
+  DRAKE_DEMAND(GetBodyIndexFromName(*tree_w_spr_, body_name) >= 0);
+  DRAKE_DEMAND(GetBodyIndexFromName(*tree_wo_spr_, body_name) >= 0);
+  body_index_w_spr_.push_back(GetBodyIndexFromName(*tree_w_spr_, body_name));
+  body_index_wo_spr_.push_back(GetBodyIndexFromName(*tree_wo_spr_, body_name));
+  pt_on_body_.push_back(pt_on_body);
+}
+void TransTaskSpaceTrackingData::AddStateAndPointToTrack(int state,
+    std::string body_name, Vector3d pt_on_body) {
+  state_.push_back(state);
+  AddPointToTrack(body_name, pt_on_body);
+}
+
+void TransTaskSpaceTrackingData::UpdateYAndError(const VectorXd& x_w_spr,
+    KinematicsCache<double>& cache_w_spr) {
+  y_ = tree_w_spr_->transformPoints(cache_w_spr, pt_on_body_.at(GetStateIdx()),
+      body_index_w_spr_.at(GetStateIdx()), 0);
+  error_y_ = y_des_ - y_;
+}
+void TransTaskSpaceTrackingData::UpdateYdot(const VectorXd& x_w_spr,
+    KinematicsCache<double>& cache_w_spr) {
+  dy_ = tree_w_spr_->transformPointsJacobian(cache_w_spr,
+        pt_on_body_.at(GetStateIdx()),
+        body_index_w_spr_.at(GetStateIdx()), 0, false) *
+        x_w_spr.tail(tree_w_spr_->get_num_velocities());
+}
+void TransTaskSpaceTrackingData::UpdateJ(const VectorXd& x_wo_spr,
+    KinematicsCache<double>& cache_wo_spr) {
+  J_ = tree_wo_spr_->transformPointsJacobian(cache_wo_spr,
+       pt_on_body_.at(GetStateIdx()),
+       body_index_wo_spr_.at(GetStateIdx()), 0, false);
+}
+void TransTaskSpaceTrackingData::UpdateJdotV(const VectorXd& x_wo_spr,
+    KinematicsCache<double>& cache_wo_spr) {
+  JdotV_ = tree_wo_spr_->transformPointsJacobianDotTimesV(cache_wo_spr,
+           pt_on_body_.at(GetStateIdx()),
+           body_index_wo_spr_.at(GetStateIdx()), 0);
+}
+
+void TransTaskSpaceTrackingData::CheckDerivedOscTrackingData() {
+  if (body_index_w_spr_.empty()) {
+    body_index_w_spr_ = body_index_wo_spr_;
+  }
+  DRAKE_DEMAND(body_index_w_spr_.size() == pt_on_body_.size());
+  DRAKE_DEMAND(body_index_wo_spr_.size() == pt_on_body_.size());
+  DRAKE_DEMAND(state_.empty() || (state_.size() == pt_on_body_.size()));
+  if (state_.empty()) {
+    DRAKE_DEMAND(body_index_w_spr_.size() == 1);
+    DRAKE_DEMAND(body_index_wo_spr_.size() == 1);
+    DRAKE_DEMAND(pt_on_body_.size() == 1);
+  }
+}
+
+// RotTaskSpaceTrackingData ////////////////////////////////////////////////////
+RotTaskSpaceTrackingData::RotTaskSpaceTrackingData(string name,
+    int n_r,
+    MatrixXd K_p,
+    MatrixXd K_d,
+    MatrixXd W,
+    const RigidBodyTree<double>* tree_w_spr,
+    const RigidBodyTree<double>* tree_wo_spr) : TaskSpaceTrackingData(name, n_r,
+          K_p, K_d, W, tree_w_spr, tree_wo_spr) {
+}
+
+void RotTaskSpaceTrackingData::AddFrameToTrack(std::string body_name,
```

Comment:
This comment is for this TD class and others.

Won't everything get out of sync, if you were to call the sequence below?
```
AddFrameToTrack(name1, pose1);
AddStateAndFrameToTrack(state, name2, pose2);
```
I don't think this code will properly assign the first frame to all FSM states, and the second to only one.

Reply:
At most one point (of the body) can follow the desired trajectory, so `state_` elements can not repeat, and the length of `state_` must be the same as `frame_pose_`'s if `state_` is not empty.

There is a method `CheckDerivedOscTrackingData` that checks if the `state_` length is correct.

I just added comments about this, so it's more obvious for the users.

Reply:
Ok for now, but will add an issue.